### ISSUE TITLE
Puzzle about MMD

### ISSUE THREAD

**Issue by Xavierxhq** [jindongwang/transferlearning#43]
thanks again for the nice program u provide. but i'm just puzzling about the mmd, here is my question: **does the ground truth labels of the datasets affects the mmd value?**
it's saying, if we have datasets A and B here, then we can compute the mmd value between A and B, but what i dont know is that: **whether the labels of the samples is used to compute the mmd value?** Cause from what i see in the program, it seems that the labels dont matter the result value.
thanks to everyone whom would check this and help me solving this puzzle.

**Comment by jindongwang** [comment]
@Xavierxhq Labels are not used in the computing process since MMD is an unsupervised method. We just use MMD to get the domain distance.

**Comment by guizilaile23** [comment]
Thanks a lot for your excellent code, and here I get a question, based on your blog about DAN(https://zhuanlan.zhihu.com/p/27657910), there is a weight parameter β for the multi-kernel need to be learned, but i could not recognize it in the DAN code you provided. is it 'fix_sigma'?

def mmd_linear(f_of_X, f_of_Y):
    delta = f_of_X - f_of_Y
    loss = torch.mean(torch.mm(delta, torch.transpose(delta, 0, 1)))
    return loss

def guassian_kernel(source, target, kernel_mul=2.0, kernel_num=5, fix_sigma=None):
    n_samples = int(source.size()[0])+int(target.size()[0])
    total = torch.cat([source, target], dim=0)
    total0 = total.unsqueeze(0).expand(int(total.size(0)), int(total.size(0)), int(total.size(1)))
    total1 = total.unsqueeze(1).expand(int(total.size(0)), int(total.size(0)), int(total.size(1)))
    L2_distance = ((total0-total1)**2).sum(2)
    if fix_sigma:
        bandwidth = fix_sigma
    else:
        bandwidth = torch.sum(L2_distance.data) / (n_samples**2-n_samples)
    bandwidth /= kernel_mul ** (kernel_num // 2)
    bandwidth_list = [bandwidth * (kernel_mul**i) for i in range(kernel_num)]
    kernel_val = [torch.exp(-L2_distance / bandwidth_temp) for bandwidth_temp in bandwidth_list]
    return sum(kernel_val)#/len(kernel_val)

**Comment by Xavierxhq** [comment]
@jindongwang Thanks a lot for the reply!

**Comment by jindongwang** [comment]
@guizilaile23 Yes. The code is a little bit different from the ones in DAN. But the ideas are the same.

**Comment by Xavierxhq** [comment]
@guizilaile23 no big deal, in this way i can get more from your question, which i should thank for.

**Comment by ChengYeung1222** [comment]
> Thanks a lot for your excellent code, it quit easy to understand compare with the code provided by official~~~~
> and here I get a question, based on your blog about DAN(https://zhuanlan.zhihu.com/p/27657910), there is a weight parameter β for the multi-kernel need to be learned, but i could not recognize it in the DAN code you provided. is it 'fix_sigma'?
> 
> def mmd_linear(f_of_X, f_of_Y):
> delta = f_of_X - f_of_Y
> loss = torch.mean(torch.mm(delta, torch.transpose(delta, 0, 1)))
> return loss
> 
> def guassian_kernel(source, target, kernel_mul=2.0, kernel_num=5, fix_sigma=None):
> n_samples = int(source.size()[0])+int(target.size()[0])
> total = torch.cat([source, target], dim=0)
> total0 = total.unsqueeze(0).expand(int(total.size(0)), int(total.size(0)), int(total.size(1)))
> total1 = total.unsqueeze(1).expand(int(total.size(0)), int(total.size(0)), int(total.size(1)))
> L2_distance = ((total0-total1)**2).sum(2) if fix_sigma: bandwidth = fix_sigma else: bandwidth = torch.sum(L2_distance.data) / (n_samples**2-n_samples)
> bandwidth /= kernel_mul ** (kernel_num // 2)
> bandwidth_list = [bandwidth * (kernel_mul**i) for i in range(kernel_num)]
> kernel_val = [torch.exp(-L2_distance / bandwidth_temp) for bandwidth_temp in bandwidth_list]
> return sum(kernel_val)#/len(kernel_val)

The bandwidth indicate gamma(the denominator inside the exponential expression)according to the paper. This is how I understand. And there's no reflect of coefficients beta as in Gretton's paper. Just add these kernels up.
